File: Online Judges/Huxley/427PaisesEmGuerra.py
from heapq import *
inf = 2**33

def floydWarshall(graph):
    dist = graph
    size = len(graph)
    for k in range(size):
        for i in range(size):
            for j in range(size):
                if (dist[i][k] + dist[k][j] < dist[i][j]):
                    dist[i][j] = dist[i][k] + dist[k][j]
    return(dist)

run = 0
while (True):
    cities, connections = map(int, input().split())
    if (cities == 0 and connections == 0):
        break
    if (run > 0):
        print()

    # dp = [[-1] * (cities + 1) for i in range(cities + 1)]
    graph = [[inf] * (cities + 1) for i in range(cities + 1)]
    for i in range(connections):
        u, v, c = map(int, input().split())
        if (graph[v][u] != inf):
            graph[u][v] = 0
            graph[v][u] = 0
        else:
            graph[u][v] = c

    dist = floydWarshall(graph)
    queries = int(input())
    for i in range(queries):
        begin, end = map(int, input().split())
        if (dist[begin][end] == inf):
            print("Nao e possivel entregar a carta")
        else:
            print(dist[begin][end])
    run += 1

File: Online Judges/Huxley/427PaisesEmGuerra.cpp
#include <bits/stdc++.h>
int inf = INT_MAX / 3;

int **floydWarshall(int **graph, int size)
{
  int **dist = (int**) malloc((size + 1) * sizeof(int*));
  for (int i = 0; i <= size; i ++)
  {
    dist[i] = (int*) malloc((size + 1) * sizeof(int));
    for (int j = 0; j <= size; j ++)
      dist[i][j] = graph[i][j];
  }

  for (int k = 0; k <= size; k ++)
    for (int i = 0; i <= size; i ++)
      for (int j = 0; j <= size; j ++)
        if (dist[i][k] + dist[k][j] < dist[i][j])
          dist[i][j] = dist[i][k] + dist[k][j];

  return(dist);
}

int main()
{
  int run = 0, cities, connections;
  while (scanf("%d %d", &cities, &connections) && !(cities == 0 && connections == 0))
  {
    if (run > 0) printf("\n");

    int **graph = (int**) malloc((cities + 1) * sizeof(int*));
    for (int i = 0; i <= cities; i ++)
    {
      graph[i] = (int*) malloc((cities + 1) * sizeof(int));
      for (int j = 0; j <= cities; j ++)
        graph[i][j] = inf;
    }

    int u, v, c;
    for (int i = 0; i < connections; i ++)
    {
      scanf("%d %d %d", &u, &v, &c);
      if (graph[v][u] != inf)
      {
        graph[u][v] = 0;
        graph[v][u] = 0;
      }
      else
        graph[u][v] = c;
    }

    int **dist = floydWarshall(graph, cities);
    int queries; scanf("%d", &queries);
    int begin, end;
    for (int i = 0; i < queries; i ++)
    {
      scanf("%d %d", &begin, &end);
      if (dist[begin][end] == inf)
        printf("Nao e possivel entregar a carta\n");
      else
        printf("%d\n", dist[begin][end]);
    }

    run ++;
    free(graph); free(dist);
  }
  return(0);
}